crypto/sha/asm/keccak1600-mmx.pl

```diff
@@ -256,13 +253,16 @@ ()
 	&psllq	(@T[1],$rhotates[4][($x+4)%5]);
 	&por	(@C[4],@T[1]);		# C[4] = ROL64(A[4][2] ^ D[2], rhotates[4][2]);
 }
-	&Rho	(3);	&Chi	(1);
-	&Rho	(1);	&Chi	(2);
-	&Rho	(4);	&Chi	(3);
+	&Rho	(3);	&Chi	(1, 1);
+	&Rho	(1);	&Chi	(2, 4);
+	&Rho	(4);	&Chi	(3, 2);
 	&Rho	(2);	###&Chi	(4);
 
 	&movq	(@T[0],@C[0]);		######### last Chi(4) is special
+	 &xor	("edi","esi");		# &xchg	("esi","edi");
 	&movq	(&QWP(@D[1],"esp"),@C[1]);
+	 &xor	("esi","edi");
+	 &xor	("edi","esi");
```

Comment:
As it turns out xchg is so-called MSROM instruction on Silvermont. And MSROM is kind of worst thing than can happen, because it stalls whole pipeline for few cycles and depending on which decoder port it ended up on it can incur additional penalty in top of "normal" overhead. In other words if MSROM instructions are inevitable, it takes careful scheduling, but best is if you can avoid them altogether. Of course question is how does alternative affect performance on *other* processors. And even more specifically by how much. It has to be by little. Like in this case. Losses on other processors are portion of percent (actually *none* on high-end processors, because not all resources are fully utilized), while on Silvermont it's several percent gain. So to answer the question, both yes and no. It's a trade-off solution, but it's weighed and conscious one.